src/fluree/db/query/fql.cljc

```diff
@@ -11,13 +15,198 @@
             [fluree.db.query.json-ld.response :as json-ld-resp]
             [fluree.db.dbproto :as db-proto]
             [fluree.db.constants :as const])
-  (:refer-clojure :exclude [vswap!])
+  (:refer-clojure :exclude [var? vswap!])
   #?(:cljs (:require-macros [clojure.core])))
 
 #?(:clj (set! *warn-on-reflection* true))
 
 (declare query)
 
+(s/def ::limit pos-int?)
+
+(s/def ::offset nat-int?)
+
+(s/def ::fuel number?)
+
+(s/def ::depth nat-int?)
+
+(s/def ::prettyPrint boolean?)
+
+(s/def ::parseJSON boolean?)
+
+(s/def ::opts (s/keys :opt-un [::parseJSON ::prettyPrint]))
+
+(defn fn-string?
+  [x]
+  (and (string? x)
+       (re-matches #"^\(.+\)$" x)))
+
+(defn fn-list?
+  [x]
+  (and (list? x)
+       (-> x first symbol?)))
+
+(s/def ::function (s/or :string fn-string?, :list fn-list?))
+
+(s/def ::filter (s/coll-of ::function))
+
+(defn wildcard?
+  [x]
+  (#{"*" :* '*} x))
+
+(s/def ::wildcard wildcard?)
+
+(defn var?
+  [x]
+  (and (or (string? x) (symbol? x) (keyword? x))
+       (-> x name first (= \?))))
+
+(s/def ::var var?)
+
+(s/def ::ref (s/or :keyword keyword?
+                   :string string?))
+
+(s/def ::selector
+  (s/or :aggregate ::function
+        :var       ::var
+        :wildcard  ::wildcard
+        :pred      ::ref
+        :map       (s/map-of (s/or :var      ::var
+                                   :wildcard ::wildcard
+                                   :ref     ::ref)
+                             ::select
+                             :count 1)))
+
+(s/def ::select (s/or :selector   ::selector
+                      :collection (s/coll-of ::selector)))
+
+(s/def ::selectOne ::select)
+
+(s/def ::selectDistinct ::select)
+
+(s/def ::selectReduced ::select)
+
+(defn asc?
+  [x]
+  (boolean (#{'asc "asc" :asc} x)))
+
+(defn desc?
+  [x]
+  (boolean (#{'desc "desc" :desc} x)))
+
+(s/def ::direction (s/or :asc asc?, :desc desc?))
+
+(s/def ::ordering (s/or :scalar ::var
+                        :vector (s/cat :direction ::direction
+                                       :field     ::var)))
+
+(s/def ::orderBy (s/or :clause     ::ordering
+                       :collection (s/coll-of ::ordering)))
+
+(s/def ::groupBy (s/or :clause     ::var
+                       :collection (s/coll-of ::var)))
+
+(def first-key
```

Comment:
Does this rely on small maps being array-maps? I'm trying to figure out if this is deterministic.

Reply:
I don't assume it's deterministic for maps with more than one entry, but I also do a [check to ensure that this map also has only one entry](https://github.com/fluree/db/pull/254/files#diff-305ee07335574d5476e0a2117e5895bc63b2086994b9e764dde30545c7c2c351R141) in the spec definition that uses this.